\documentclass[10pt]{article}
\usepackage[english]{babel}
\input{common-headers}

\title{Mastering the Vidjil web application:\\ Browsing and analysing clones
}

\begin{document}

\maketitle

\textit{The goal of this practical session is to learn
common ways to visualize, filter, analyse and cluster clones
on the Vidjil web application.
%
These clones may have been computed by the Vidjil algorithm
or by any other algorithm.}

\bigskip

\question{Connect to the public server (\url{https://app.vidjil.org}), either with your account
or the demo account (\texttt{demo@vidjil.org} / \texttt{demo}),
select the \textit{Demo LIL-L3 (tutorial)} patient, and click on the bottom right link, \com{see results: multi-inc-xxx}.
Do not open the \textit{Demo LIL-L3 (analyzed)} patient: this one contains the complete
analysis.
The Vidjil web application opens.}

This patient (patient 063 from \href{http://dx.doi.org/10.1016/j.leukres.2016.11com.009}{Lille study on the feasibility of MRD using HTS}) suffering T-ALL has one diagnosis sample,
with dominant clones both in IGH and TRG,
and four follow-up samples, including a relapse.

\question{In the \com{settings} menu, try the various options for \com{sample key}.
  The five samples can be labeled by their name, their date of sampling or by the number of days after the first sample.
}

In the following sections, we focus on the diagnosis sample.
% and explore how to assess the quality of the data
% how to view and filter clones and
The section~\ref{sec:tracking} will deal with the comparison of several samples. 

\section{Assessing the quality of the run and of the analysis}
The Vidjil web application allows to run several ``RepSeq'' (immune repertoire analysis) algorithms.
Each RepSeq algorithm has its own definition of what a clone is (or, more precisely
a clonotype), how to output its sequence and how to assign a V(D)J designation.
The number of analyzed reads will depend on the algorithm used.
This sample has been processed using the Vidjil algorithm.


\marginpar{The percentage of analyzed reads can range from .01\,\% (for
  RNA-Seq or capture data) to 98-99\,\% (for very high-quality runs mostly on
  Illumina).}
\question{How many reads have been analyzed in the current sample with the embedded algorithm ?}

Now we will try to assess the reason why some reads were not analyzed in our
sample.
This might reflect a problem during the sequencing protocol\dots or that could
be normal.
For that sake you will need to display the information box by clicking on the
\textit{i} in the upper left part.
\question{What are the average read lengths on IGH? and on TRG?}
The lines starting with \texttt{UNSEG} display the reasons why some reads have
not been analyzed.

You can see what those reasons mean in the online documentation of the
algorithm: \href{http://www.vidjil.org/doc/vidjil-algo\#unsegmentation-causes}{vidjil.org/doc/vidjil-algo\#unsegmentation-causes
}
\question{What are the major causes explaining the reads have not been
  analyzed? Also have a look at the average read lengths of these causes. Do
  you notice something regarding the average read lengths?}

\input{tutorial-clones}

\bigskip

\bigskip

\centerline{\includegraphics[width=.5\textwidth]{./grid-1}}

\newpage


\input{tutorial-several-samples}
\input{tutorial-export}

\input{tutorial-server}

\vfill
\flushright \it Aurélien Béliard, Aurélie Caillault, Mathieu Giraud, Tatiana Rocher, Mikaël Salson, Florian Thonier
\\ \texttt{contact@vidjil.org}

\end{document}
