\section{Viewing and filtering clones}


\subsection{Looking at a clone}

Each RepSeq algorithm has its own definition of what a clone is (or, more precisely
a clonotype), and on how to output its sequence and how to assign a V(D)J designation.

In this file, the most abundant clone
is \texttt{IGHV3-9 7/CCCGGA/17 J6*02}.

\question{Select this clone, either by clicking on the list or on the grid.
  How many reads do this clone represent? (see again the bottom part to the right)}

\textcolor{white}{ the bottom panel display information about currently selected clones -> 189 991 reads (9.665\,\%) }

There are several options to display the V(D)J designation.

\question{In the  \com{settings} menu, under \com{N regions in clone names} select \com{length} to show N zones by their length. Revert to the
  default \com{sequence (when short)} setting to show the full N on short sequences.}
  
\question{Try also the options  \com{alleles in clone names} : by selecting \com{always}, the clone
  V gene is displayed as \com{IGHV3-9*01}. Revert to the default \com{when not
    *01}. This setting, which is the default, allows to have a more condensed
  V(D)J designation that doesn't make the \com{*01} appear (it is implicit).}

\subsection{Showing more clones}

By default Vidjil displays the 50 most abundant clones at each time point.
With five time points, we may therefore have from 50 to 250 clones displayed
depending if the top 50 are always the same or always different or, more
realistically, in-between.
This number can be increased to a maximum of 100 clones by going to the \com{filter} menu and by putting the
slider to its right end.
\question{Notice how the IGH smaller clones percentage (second clone displayed in the list) changes. What was its
  initial value? What is it now?}

\textcolor{white}{ filter set to 50  -> IGH smaller clones 10.11\,\%\\*
 filter set to 100  -> IGH smaller clones 8.92\,\%\\*}

The \textit{smaller clones} correspond to clones that are not displayed
because they are never among the most abundant ones.


\subsection{Tagging and filtering clones}

Consider the most abundant clones in the list:  \texttt{IGHV3-9 7/CCCGGA/17 J6*02} and  \texttt{TRGV10 13//5 JP1}.
Usually we may want to tag them in order to remember them later on.
\question{Click on the star and choose colored tags for these two clones, such as \texttt{clone 1} or \texttt{clone 2}.
  Notice how the color applies throughout all the views.}

Later you may want to filter clones depending on the tags you have chosen.

\question{In the upper left part, click on the little dark gray square (the
  second coloured square starting from the right). What happens? What if you click again?}

This is a way of filtering some clones. This may be useful when we want to
focus on some specific clones. Another way of doing so is to filter them by
their gene names or by their DNA sequences.  
\question{In the search box,
  enter \texttt{GGAGTCGGGG} and validate with \texttt{Enter}.  How many sequences are
  left?}
Note that the search is performed both on the forward and the reverse strand.
\question{Check that by searching for the reverse complement of the
  sequence: \texttt{CCCCGACTCC}. Do you find the same results as previously?}
\question{How can you cancel this filter and view again all the clones?}

\bigskip

Another solution to tag a specific clone is to rename it.
\question{Double click on the name of a clone (in the list of clones) and
  choose another name (\textit{e.g.} interesting clone) and validate using
  \texttt{Enter}.}


\bigskip

After this rename, you can see that the clone is still selected.
\question{Click on several clones by holding the \texttt{Ctrl} key to select
  more. Each time you add a new clone to the selection, its sequence
  is added in the bottom part.}

\question{How many clones are selected? How many reads do those clones
  represent?}

\question{ Notice the star at the the right of the screen, near the number
  of reads. You can also tag clones using this icon. In that way, you will be able to tag
  all the selected clones at once.}

\question{When you want to focus on the selected clones, you can click on the
  focus link on the right, next to the number of selected clones.
  This feature is useful when you want to analyse some clones more thoroughly
  without being annoyed by other clones.}

\question{To remove this focus, click on the cross next to the search box,
  above the list.}

\question{To unselect them all, you can click in an empty area on the top or
  bottom plot.}

Sometimes, one wants to hide noisy or unrelated clones.

\question{Select a clone or several clones and click on the \com{hide} button, near the \com{focus} button. Show again these
  clones by clicking on the cross next to the search box.}

% Another way to hide clonesis to assign is to change the tag of it as ``standard (niose)`` and choose to uncheck this tag by clicking on the corresponding tile on the list of tiles at the informatons panel to switch them from a visible state to a filter one.
%%% Voir ci-dessus, déjà mis

It is also possible to filter samples that do not contain a clone.  When you
have lots of samples it helps to keep the sample of interest.  Here the number
of sample is quite limited, so the feature may appear less useful.

\question{\new Click on the \com{IGHV3-11 / IGHJ6} clone in the last sample, whose
  abundance is around 10\,\%.
  Then go in the menu at the upper-right corner of the graph (where \com{5/5}
  is written) and select \com{focus on selected clones}.
}

By selecting this, the samples where this clone doesn't appear are hidden.
This is useful for instance to assess the contamination among dozens of
samples.

\question{You can go back to the previous view by returning into the menu and
  clicking on \com{show all}. Notice also how in the menu you can select the
  samples to be shown.}

\section{Analysing clone populations}

\subsection{Clustering clones through inspection of their sequences}

The first thing to be done is to see if some clones should be clustered (because
of sequencing or PCR errors for instance). This step could be automatized
but, in any case, the automatic clustering would need to be checked by an expert
eye.

By default in the bottom plot (the \textit{grid}), the clones
  are displayed according to their V and J genes (or more generally to their
  5' and 3' genes). 

\question{Identify in the grid the clones with an
  \textit{IGHV-3-13}~\textit{IGHJ6} recombination and select them
  all. You can do so either by holding \texttt{Ctrl} or by drawing a rectangle around the clones while
  maintaining down the left button of the mouse.}

The sequences of the clones now appear in the bottom part of the browser (the
\textit{sequence panel}). If many clones are selected you can view more sequences
by moving the mouse above the sequence panel.
 In such a case, you may be bothered by the sequence panel going up and
down each time your mouse enters or exits the sequence panel. You can stick it
in its current shape by clicking on the pin at the upper right corner of the
sequence panel.

Then, the sequences in the sequence panel can be visually compared but you can also align
them to see more easily their similarities.


\question{Click on the \com{align} button on the left-hand side. The differences are
emphasized in bold.}

Now it is the user's expertise to determine if sequences are sufficiently
similar, depending on her or his specific question. If some sequences don't appear to be similar enough, you can remove
them from the sequence panel by clicking on the cross in front of the sequence in
the sequence panel.
\question{Remove all the sequences that are not similar enough with the first
  one.}

Now all the sequences in the sequence panel should be highly similar. All their
differences could be due to sequencing or PCR errors.
These artifacts (mutations, homopolymers, insertions, deletions)
depend on the sequencer and the PCR technique.

\question{Cluster all those clones in a single clone by clicking on the ``cluster''
  button, next to the \com{align} button.}

All the clustered sequences now appear within a same clone. That can be seen
in the list: the clone which hosts the subclones appears with a $+$ on its
left. You can click on the $+$ to see the subclones that have been clustered in
the main one.
\question{Click on the $+$ and observe the changes in the grid.}

As you may have noticed the subclones appear again in the grid. You can
compare their sequences again if you'd like (for example to double check that
you were right to cluster them). You can also remove some subclones from the
cluster by clicking on the cross at their left in the list.
\question{For the sake of the exercise, remove the last clone of the cluster.}

\question{%
%For the next step, choose the preset \com{V distribution} (keyboard shortcut \com{5}).
% On n'a pas encore parlé ici des presets. 
Open the \com{cluster} menu, and choose \com{cluster by V/5}. What happened ? There are now two clones with TRGV2. Why ?}
%%  Confirm this by changing the x axis into ``V allele``.
%%% -> Problème, on n'a pas encore parlé des axes à cet endroit.

\question{In the \com{cluster} menu, select  \com{revert to previous clusters} to undo these clusterings.}


\subsection{Other metrics and analysis on the clones}

As a proxy to sequence similarity we used the V and J genes, however there are
other ways to assess sequence similarity that may be more pertinent.
Moreover you may want to plot other metrics on the lymphocyte population.
%
For instance we can choose to plot the V genes versus the length of the N
insertions.
\question{Go to the \com{plot} menu (in the upper left corner of the grid),
  and in the preset box choose \com{V/N length}.}

Then you can continue aligning and clustering clones if necessary.

\question{You can also try the preset \com{clone average read length/GC content}
  which tends to separate quite nicely the distinct clones.}

Note that you can choose any axis to be plotted: just go the \com{plot} menu and
select any value you would like for the $x$ axis and for the $y$ axis.
For bar charts, the box sizes always relates to the clone size,
and the $y$ axis selects the order of the boxes sharing a same $x$).

%% \item Regarder les stats disponibles, mettre n°7 (taille des reads)

\question{In the \com{plot} menu, switch between the ``bubble plot'' and the ``bar plot''.
In the bar plot mode, pass the mouse over the bars: What happens?}

% Another possibility is to request Vidjil to compute the similarity between
% clones.
% \question{Now select the preset \com{plot by similarity} or even \com{plot
%   similarity by locus} to plot similarity for the current locus (beware: this
% may take some time).}
% Now the most similar clones should be close together. However note that it is
% theoretically impossible to achieve such a representation in 2 dimensions. So
% it is possible that two dissimilar clones are close together or, conversely,
% that two similar clones are far apart.

\question{Press the keys \texttt{0} to \texttt{9} on the numeric keypad. What happens ?}

There is still a feature to help you analyse your data that we have not
explored yet.
You can change the colors to make it represent some variables of interest
with the \com{color by} menu.
\question{First choose the preset \com{V/J (genes)} and
  then color by \com{N length} (in the box at the top of the screen).}
\marginpar{We apologize to color blinds: the colors are not yet color-blind friendly.}Clones that are close on the grid with similar colors are likely to
be similar.

\question{Choose now the preset \com{CDR3 length distribution} and
  then color by \com{productivity}.
  See that the color tiles in the info part (upper right) change to show the color key.}

\question{ Instead of coloring by productivity, you could also color by
  \com{clone}. When coloring by \com{clone}, each clone has a random color. Thus in
  a bar plot, it is a convenient color mode to see the peaks that are due to a
  single clone or to several clones.
  However clones may be very similar. Another option is to color by
  \com{CDR3}. In such case all clones with the same CDR3 will have the same
  color (note that, due to a lack of available colors two different CDR3s
  could share the same color just by chance).}


Using those different features you should be able to analyse how similar your
sequences are, and potentially you could cluster them if you'd like or tag them.

\question{
  Select the most abundant clone. It now appears in the sequence panel.
  Now we would like to compare the sequence with the germline genes.
  We can add the germline genes to the sequence panel by going
  to the \com{import/export} menu and by clicking on \com{add germline genes}.
  Now we can click on the \com{align} button to see the alignment between the
  genes and the sequence. Mutations can be identified and silent mutations are
  displayed with a double border in blue.
}

\bigskip

\textit{This part is specific to samples analyzed with Vidjil-algo.}

Some clones may be less trustable than other ones\dots{} Let's see how to spot them.
\question{In the clone list, search clones with an orange warning at the
  right side. Click on the warning. What are the warnings due to?}

There may have several reasons: 
\begin{itemize}
\item average coverage: in that case the clonal sequence displayed is short
  compared to the reads in the clone. This may be the case when too different
  sequences have been put in a clone. The value is generally $\geq 80\,\%$.
\item $e$-value: It is a statistical value computed to ensure that
  recombinations have not been spot by chance. This value is generally much
  lower than 1 ($<10^{-5}$).
\item Clone similar to another one: In that case Vidjil-algo tells you that
  other clones have the same genes and may be similar
\item Non-recombined sequences: Some known unrecombined sequences are tagged
  so that you can spot them easily. We tag the unrecombined IGHD7-27/IGHJ1
  sequences that may be amplified.
\end{itemize}

You can view those values for any clone by clicking the \textit{i} icon on the
right side, in the list of clones.
\subsection{Analysing recombinations from several loci}

First make sure to come back to the preset \com{V/J (genes)} in the \com{plot} menu.

If you want to focus on specific locus, you can click on the locus name in
the upper left part. One click will make the locus disappear, another one will
make it appear again.
If you hold the \texttt{Shift} key (the one which is usually above the left
\texttt{Ctrl} key) while clicking it will hide all the loci but the one you
clicked on.

\question{Click on \com{IGH}, while holding the \texttt{Shift} key. Now what is the
  number of analyzed reads? Why did it change?}

\question{Now click on \com{TRG}, to filter it in again.}

\question{Press on the \texttt{g} key. What happens? Now, press on the
  \texttt{h} key. Press on the \texttt{g} again (you can do that anytime you
  like :)). Let's stick to the TRG locus.}

You can also change the current locus by clicking on the locus name in the
right part of the grid.


\subsection{Clone quantification (using spike-ins)}

Sometimes you may include spike-ins in your sample to allow a more reliable
quantification.
Let us assume that the main clone with IGHV-3-9 / IGHJ5 is a spike-in whose
expected concentration is 1\% (.01).

\question{First let's color this clone with the \com{standard} tag.}

\question{ Now we will set its concentration to .01 as expected. Click again on
  the star. In the \com{normalize to} field enter \com{.01} and click \com{ok}.
  Now, in the graph, this clone should correspond to a straight line at 1\%.}

\question{ Notice how the concentrations of the other clones have changed
  accordingly.
  You can go to the \com{settings} menu to disable this normalization and to
  go back to the raw concentrations.}

Then you can set expected concentrations for other clones and you are free to
switch between those normalizations.
It is also possible to set up normalization against external data,
contact us if you are interested.

