
\section{Assessing the quality of the run and of the analysis}
The Vidjil web application allows to run several ``RepSeq'' (immune repertoire analysis) algorithms.
Each RepSeq algorithm has its own definition of what a clone is (or, more precisely
a clonotype), how to output its sequence and how to assign a V(D)J designation.
The number of analyzed reads will depend on the algorithm used.
This sample has been processed using the Vidjil algorithm.


\marginpar{The percentage of analyzed reads can range from .01\,\% (for
  RNA-Seq or capture data) to 98-99\,\% (for very high-quality runs mostly on
  Illumina).}
\question{How many reads have been analyzed in the current sample with the embedded algorithm ?}

\reponse{In the upper left corner, you can see an information panel with \com{analyzed reads 1 967 338 (82.31\,\%)'}}

Now we will try to assess the reason why some reads were not analyzed in our
sample.
This might reflect a problem during the sequencing protocol\dots or that could
be normal.
For that sake you will need to display the information box by clicking on the
\textit{i} in the upper left part.
\question{What are the average read lengths on IGH? and on TRG?}

\reponse{In the Analysis log row, under \com{av. len}\\*
IGH \fl 314.5\\*
TRG \fl 197.6 }

The lines starting with \texttt{UNSEG} display the reasons why some reads have
not been analyzed.

You can see what those reasons mean in the online documentation of the
algorithm:

 \centerline{\href{http://www.vidjil.org/doc/vidjil-algo/\#reads-without-detected-recombinations}{vidjil.org/doc/vidjil-algo\#reads-without-detected-recombinations}}

\question{What are the major causes explaining the reads have not been
  analyzed? Also have a look at the average read lengths of these causes. Do
  you notice something regarding the average read lengths?}


\reponse{ 1. The algorythm was not able to find a V or a J for most of the unsegmeneted reads.\\*
2. The may be too short to cover enough of the V or J genes to be detected. }
